kimchi/src/circuits/gate.rs

```diff
@@ -97,17 +97,17 @@ pub enum GateType {
 #[serde_as]
 #[derive(Clone, Debug, Serialize, Deserialize)]
 /// A single gate in a circuit.
-pub struct CircuitGate<F: FftField> {
+pub struct CircuitGate<G: KimchiCurve> {
```

Comment:
why do we have to change to KimchiCurve here btw?

Reply:
It is needed for `CircuitGate::verify()`

Reply:
1. I considered that, but equality constraints in where clauses seems to be unsupported.
2. The problem is that the methods that no longer need the `ConstraintSystem` now need access to the curve generic, like `verify_poseidon()`, I could only remove the constraint system as parameter because it can now access the same information from the curve generic. Also range check gate seems to need the constraint system for things that cant be provided from the generic.

The only way I see is to move `verify` out of the struct.

Reply:
> The only way I see is to move verify out of the struct.

That would work, or since that verify function is actually called from `ConstraintSystem::verify` which knows about the parameters + endo values, it could pass these values down to `CircuitGate::verify` as arguments.

Reply:
I just tried moving it out of `CircuitGate`, the problem is that `verify()` calls other functions that are also part of `CircuitGate`.
But I remember that there is another way to do the same as 1 that is supported.
Instead of :
`verify<G>(...) where G: KimchiCurve, G::ScalarField = F`
In this case is possible to do:
`verify<G>(...) where G: KimchiCurve<ScalarField = F>`
So that will allow the few functions that need it to access the curve while `CircuitGate` is generic only over the field, also in a more general way this allows to have `F` as a kind of alias for `G::ScalarField`.

Reply:
This latter idea would be nice since right now some parts of the code related to the types are less readable since we went from `F` to `G::ScalarField`. If we could stick to `F` it would be easier.
In any case, I agree with @mimoo that keeping structs generic over fields seems more intuitive than over curves.